docs: Rewrite Settings and Workspaces documentation (#9065)
iTwin/itwinjs-core: docs/learning/backend/Settings.md

> # Settings

Reviewer: Really thorough document here, @hl662, nice work. I would suggest giving users a bit more up front in the introductory paragraphs. Otherwise, readers won't get an idea of things like priority or the fact that there are different levels and locations of settings until they come across way down the page.

I wouldn't go into detail, but a sentence about how they operate at different levels of persistence and have different priorities could be helpful to know up front.

Author: Good suggestion — I added a short sentence up front that settings can live at different persistence levels and are resolved by priority, so readers get that framing earlier. Addressed in d92fc82919.

[Settings]($backend) are how administrators configure an iTwin.js application for end-users. A setting is a named value — a `string`, `number`, `boolean`, object, or array — that controls application behavior at run-time.

> **New to this topic?** Start with the [Workspaces and Settings overview](./WorkspacesAndSettings.md) to understand how [Settings]($backend), settings containers, and [WorkspaceDb]($backend) relate before diving in here.

## What is a setting?

> Be careful to avoid confusing **settings** with **user preferences**, which are configured by individual users. For example, an application might provide a check box to toggle "dark mode". Each individual user can make their own choice — it is a user preference, not a setting. But an administrator may define a setting that controls whether users can see that check box in the first place.

Reviewer: A bit weird that the first sentences in the "What is a setting?" section are about what a setting is _not_. I might move this paragraph down below, after settings are introduced a bit more. Or rephrase it, so you lead with the example dark mode checkbox, then mention the choice/value is a user preference.

Author: Agreed — I rewrote that section to lead with what a setting does, then contrasted it with user preferences using the dark mode example. Addressed in d92fc82919.

```ts
[[include:WorkspaceExamples.RegisterSchema]]
```

Register schemas shortly after invoking [IModelHost.startup]($backend), using the [IModelHost.onAfterStartup]($backend) event:

```ts
IModelHost.onAfterStartup.addListener(() => {
  IModelHost.settingsSchemas.addGroup(landscapeProSchema);
});
```

All schemas are unregistered when [IModelHost.shutdown]($backend) is invoked.

### Schema validation behavior

Validation occurs lazily when you **retrieve** a setting value — not when the value is stored. When you call methods like [Settings.getString]($backend), [Settings.getObject]($backend), or [Settings.getBoolean]($backend), the value is validated against the registered schema for that setting name:

- If **no schema** is registered for the setting name, the value passes through unchecked.
- If a schema **is** registered, type mismatches throw an error — for example, if the schema declares a setting to be a `string` but the dictionary supplies a `number`.
- For `object` types, `required` fields are enforced and `extends` references are resolved recursively.

> Because validation happens on retrieval, register your schemas early — ideally via [IModelHost.onAfterStartup]($backend) — so that type errors are caught as soon as the setting is first accessed.

Reviewer: Slightly unclear why you wouldn't register in `iModelHost.onAfterStartup`. Several lines above, you tell users explicitly to "Register schemas shortly after invoking [IModelHost.startup]($backend), using the [IModelHost.onAfterStartup]($backend) event"

Author: Good catch — I kept the recommendation to register in `IModelHost.onAfterStartup`, and added a note that some later extract snippets call `addGroup(...)` directly only for brevity. Addressed in d92fc82919.

## Settings dictionaries

> The values of [Setting]($backend)s are provided by [SettingsDictionary]($backend)s. The [Settings]($backend) for the current session can be accessed via the `settings` property of [IModelHost.appWorkspace]($backend).

Reviewer: ```suggestion
Values of [Setting]($backend) are provided by [SettingsDictionary]($backend). The [Settings]($backend) for the current session can be accessed via the `settings` property of [IModelHost.appWorkspace]($backend).
```

I think it reads alright without the 's'? A pet peeve of mine.

Author: Agreed — I rewrote this rather than taking the suggestion verbatim. It now says a `SettingsDictionary` supplies values for one or more `Setting`s. Addressed in d92fc82919.

Any number of dictionaries can be added to [Workspace.settings]($backend). Let's add another:

```ts
[[include:WorkspaceExamples.AddSecondDictionary]]
```

This dictionary adds a value for "landscapePro/flora/preferredStyle", and defines new values for two settings already defined in the previous dictionary. When we look up those settings again:

```ts
[[include:WorkspaceExamples.GetMergedSettings]]
```

"landscapePro/flora/preferredStyle" is no longer `undefined`. The value of "landscapePro/ui/defaultTool" has been overwritten. And "landscapePro/ui/availableTools" contains the merged contents of both dictionaries — because the `settingDef` has [SettingSchema.combineArray]($backend) set to `true`.

### Settings priorities

Configurations are often layered. [SettingsPriority]($backend) defines which dictionaries take precedence — the highest-priority dictionary that provides a value for a given setting name wins.

Specific values:

- [SettingsPriority.defaults]($backend) (100) — settings loaded automatically at startup.
- [SettingsPriority.application]($backend) (200) — settings supplied by the application to override or supplement defaults.

> - [SettingsPriority.organization]($backend) (300) — settings applying to all iTwins in an organization.

Reviewer: "settings applied" here and below to match the tense: "settings supplied" and "settings loaded" above?

Author: Makes sense — updated these bullets to use `settings applied ...` for consistency with the surrounding phrasing. Addressed in d92fc82919.

- [SettingsPriority.iTwin]($backend) (400) — settings applying to a particular iTwin and all its iModels.
- [SettingsPriority.branch]($backend) (500) — settings applying to all branches of a particular iModel.
- [SettingsPriority.iModel]($backend) (600) — settings applying to one specific iModel.

> [SettingsDictionary]($backend)s of `application` priority or lower reside in [IModelHost.appWorkspace]($backend). iTwin-scoped settings are loaded into the workspace returned by [IModelHost.getITwinWorkspace]($backend) — see [iTwin settings](#itwin-settings). Settings of even higher priority (branch and iModel) are stored in an [IModelDb.workspace]($backend) — see [iModel settings](#imodel-settings).

Reviewer: ```suggestion
A [SettingsDictionary]($backend) of `application` priority or lower resides in [IModelHost.appWorkspace]($backend). iTwin-scoped settings are loaded into the workspace returned by [IModelHost.getITwinWorkspace]($backend) — see [iTwin settings](#itwin-settings). Settings of even higher priority (branch and iModel) are stored in an [IModelDb.workspace]($backend) — see [iModel settings](#imodel-settings).
```

Author: Agreed — switched this sentence to singular generic wording. Addressed in d92fc82919.

```ts
[[include:Settings.dropITwinDictionary]]
```

> Note: The examples above use [Settings.addDictionary]($backend), which loads dictionaries into memory for the current session only. For cloud-backed iTwin and organization settings — where dictionaries are fetched from `SettingsDb` containers on demand — see [SettingsDb](#settingsdb) below.

What about "landscapePro/ui/availableTools"? In the LandscapePro schema, the corresponding `settingDef` has [SettingSchema.combineArray]($backend) set to `true`, meaning that when multiple dictionaries provide a value for the setting, they are merged into a single array, eliminating duplicates, and sorted in descending order by dictionary priority.

### How settings are stored

The sections below cover three progressively richer ways to work with settings:

- **Session-only (in-memory):** [Settings.addDictionary]($backend) loads a dictionary for the current session. Good for testing, app defaults, and short-lived overrides.
- **iTwin / iModel persisted:** [IModelHost.saveSettingDictionary]($backend) writes named dictionaries to cloud storage scoped to an iTwin or iModel. Changes persist across sessions and are shared with other users.

> - **SettingsDb (cloud-managed storage):** A [SettingsDb]($backend) stores individual settings as key-value pairs in a dedicated container with its own versioning and access control — see [SettingsDb](#settingsdb).

Reviewer 2: I would hesitate to even mention SettingsDbs here. Every SettingsDb is linked to an iTwin anyways - how could you create a SettingsDb without implying its not iTwin persisted?

iModel-scoped SettingsDbs should be exposed in a similar manner - `IModelDb.saveSettingsDictionary` should be the only way to do that

Author: Makes sense. The Settings.md doc now treats SettingsDb as an implementation detail — it's only mentioned in the "Settings containers" section as context for how `WorkspaceEditor` works, not as something users create directly. The primary path leads with `getITwinWorkspace`.

## iTwin settings

> Each iTwin has its own workspace with its own [Settings]($backend) that can override and/or supplement the application workspace's settings. These settings are stored as named [SettingsDictionary]($backend)s in a [SettingsDb]($backend) (a settings-formatted [WorkspaceDb]($backend)) scoped to that iTwin. Whenever [IModelHost.getITwinWorkspace]($backend) is called, all named dictionaries in the container are loaded into the returned [Workspace.settings]($backend) at [SettingsPriority.iTwin]($backend). An application working in the context of a particular iTwin should resolve setting values by asking [IModelHost.getITwinWorkspace]($backend), which will fall back to [IModelHost.appWorkspace]($backend) if the iTwin's setting dictionaries don't provide a value for the requested setting.

Reviewer 2: this is the correct way to talk about settingsDbs imo. In the context of iTwin Settings - almost an implementation detail

Author: 👍 That's the framing I've adopted throughout — SettingsDb is mentioned only as the underlying persistence layer, not as a user-facing concept.

```ts
[[include:WorkspaceExamples.GetITwinWorkspace]]
```

The returned [Workspace]($backend) gives you access to all settings and resources associated with the iTwin.

### Saving iTwin settings

To save a named settings dictionary for an iTwin, call [IModelHost.saveSettingDictionary]($backend) with a dictionary name and a [SettingsContainer]($backend) of key-value pairs:

```ts
[[include:WorkspaceExamples.SaveITwinSettings]]
```

If no settings container exists for the specified iTwin yet, one is created automatically. The dictionary name becomes the resource name in the [SettingsDb]($backend). Multiple named dictionaries can coexist in the same container.

### Deleting iTwin settings

To remove an entire named dictionary, use [IModelHost.deleteSettingDictionary]($backend):

```ts
[[include:WorkspaceExamples.DeleteITwinSetting]]
```

### Reading iTwin settings

To read iTwin settings, query the `settings` of the [Workspace]($backend) returned by [IModelHost.getITwinWorkspace]($backend):

```ts
[[include:WorkspaceExamples.ReadITwinSettings]]
```

iTwin settings are loaded with [SettingsPriority.iTwin]($backend) priority.

## iModel settings

> > **⚠️ Caution — prefer SettingsDb for new work.** Storing settings directly in an iModel (via [IModelDb.saveSettingDictionary]($backend)) ties configuration to a single iModel file. Settings cannot be discovered without opening the iModel, cannot be versioned independently, and require an exclusive write lock on the entire iModel to modify. For new projects, store settings in a cloud-hosted [SettingsDb](#settingsdb) instead — it is discoverable by iTwinId, versioned independently, and does not require an iModel to be open.

Reviewer 2: "require an exclusive write lock on the entire iModel to modify" - I keep seeing this but where do we actually enforce that? I don't think this is true.

Author: Good catch — I softened this. The current text says the lock is "obtained automatically by `withEditTxn`" rather than claiming it's enforced at the settings level. If that's still overstating it, happy to remove the lock mention entirely and just say ordinary users shouldn't do this. What do you think?

> > **⚠️ Caution — prefer SettingsDb for new work.** Storing settings directly in an iModel (via [IModelDb.saveSettingDictionary]($backend)) ties configuration to a single iModel file. Settings cannot be discovered without opening the iModel, cannot be versioned independently, and require an exclusive write lock on the entire iModel to modify. For new projects, store settings in a cloud-hosted [SettingsDb](#settingsdb) instead — it is discoverable by iTwinId, versioned independently, and does not require an iModel to be open.

Reviewer 2: I wouldn't caution against iModel settings - there are legitimate use-cases for them and the only option is to store them in the iModel until we modify saveSettingsDictionary to *not* do that

Author: Fair point. I've reframed the note — it now describes the tradeoff (tied to iModel file, not discoverable without opening it, not independently versioned) without steering users away. The tone is informational rather than cautionary.

> Each [IModelDb]($backend) has its own [Workspace]($backend), accessible via [IModelDb.workspace]($backend). This workspace inherits all app-level and iTwin-level settings, and layers on settings stored inside the iModel itself. Because these iModel-level dictionaries are loaded at [SettingsPriority.iModel]($backend) — the highest built-in priority — they override any lower-priority setting with the same name.

Reviewer 2: The workspace does not inherit iTwin-level settings by-default today

Author: Fixed in both Settings.md and Workspace.md. The docs now explicitly state that `IModelDb.workspace` does not auto-include iTwin settings, and the "Referencing iTwin settings from an iModel" section shows the manual bridging pattern.

```

Now when the application reads `landscapePro/flora/preferredStyle` from this iModel's workspace scope, it gets `"formal"`. All other iModels in the iTwin continue to use the iTwin-level value.

### Referencing iTwin settings from an iModel

An iModel doesn't inherently know which iTwin settings container it should use. By saving the container's identity as an iModel-level setting, the iTwin settings become part of the iModel's workspace — so [IModelDb.workspace]($backend) becomes the single place to resolve all settings and resources the iModel uses.

```ts
[[include:WorkspaceExamples.SaveITwinSettingsReferenceInIModel]]
```

The next time the iModel is opened, your app reads `landscapePro/itwinSettingsRef` and uses it to load the same iTwin settings container. By default this resolves to the latest version of those settings.

### Pinning iTwin settings versions

If you need to pin the iModel to a specific version of the iTwin settings — so that its configuration does not change even when the iTwin settings are updated — save the version alongside the container props:

```ts
[[include:WorkspaceExamples.VersionAndPinITwinSettings]]
```

> ## SettingsDb

Reviewer 2: I wouldn't add this to the docs. Imo, a SettingsDb is the underlying persistence layer (and the first layer) to *any* workspace (today is just for the iTwin workspace). But why does a user need to know that? I don't think we want users creating SettingDbs at will (remember, there can only be *one* per iTwin and so its important its created correctly).

Author: Agreed — removed the standalone SettingsDb creation guidance. The "Settings containers" section now only mentions `WorkspaceEditor` in the context of admin tooling, with a callout that `saveSettingDictionary` handles the single-container constraint automatically.

# Settings

[Settings]($backend) are how administrators configure an iTwin.js application for end-users. A setting is a named value — a `string`, `number`, `boolean`, object, or array — that controls application behavior at run-time.

> > **New to this topic?** Start with the [Workspaces and Settings overview](./WorkspacesAndSettings.md) to understand how [Settings]($backend), settings containers, and [WorkspaceDb]($backend) relate before diving in here.

Reviewer: nice :)

## What is a setting?

Be careful to avoid confusing **settings** with **user preferences**, which are configured by individual users. For example, an application might provide a check box to toggle "dark mode". Each individual user can make their own choice — it is a user preference, not a setting. But an administrator may define a setting that controls whether users can see that check box in the first place.

A [Setting]($backend) is a name-value pair. The value can be:

- A `string`, `number`, or `boolean`;
- An `object` containing properties of any of these types; or
- An `array` containing elements of one of these types.

A [SettingName]($backend) must be unique, 1 to 1024 characters long with no leading nor trailing whitespace, and should begin with the schema prefix of the [schema](#settings-schemas) that defines it. For example, LandscapePro might define:

```
  "landscapePro/ui/defaultToolId"
  "landscapePro/ui/availableTools"
  "landscapePro/flora/preferredStyle"
  "landscapePro/flora/treeDbs"
  "landscapePro/hardinessRange"
```

> Each name begins with the "landscapePro" schema prefix followed by a forward slash. Forward slashes create logical groupings, similar to file paths.

Reviewer: ```suggestion
Forward slashes create logical groupings, similar to file paths.
```

```

"landscapePro/flora/preferredStyle" is no longer `undefined`. The value of "landscapePro/ui/defaultTool" has been overwritten. And "landscapePro/ui/availableTools" contains the merged contents of both dictionaries — because the `settingDef` has [SettingSchema.combineArray]($backend) set to `true`.

### Settings priorities

Configurations are often layered. [SettingsPriority]($backend) defines which dictionaries take precedence — the highest-priority dictionary that provides a value for a given setting name wins.

```mermaid
graph TD
    subgraph AppWorkspace["IModelHost.appWorkspace"]
        D["defaults (100)<br/><i>files loaded at startup</i>"]
        A["application (200)<br/><i>supplied by the app at run-time</i>"]
        O["organization / iTwin (400)<br/><i>settings loaded for the active iTwin</i>"]
    end

    subgraph IModelWorkspace["IModelDb.workspace"]
        B["branch (500)"]
        M["iModel (600)<br/><i>settings loaded for the active iModel</i>"]
    end

    D -->|"overridden by"| A
    A -->|"overridden by"| O
    O -->|"overridden by"| B
    B -->|"overridden by"| M

    style M fill:#d4edda,stroke:#28a745
    style D fill:#f8f9fa,stroke:#6c757d

> ```

Reviewer 2: I would omit this mermaid diagram. Its problematic because it says iTwin settings are in IModelHost.appWorkspace which isn't true, but also I just don't see the value beyond what you already explained

Author: Agreed, removed the diagram. The bullet list above it covers the same ground.

Each [IModelDb]($backend) has its own [Workspace]($backend), accessible via [IModelDb.workspace]($backend). This workspace inherits app-level settings (via [IModelHost.appWorkspace]($backend)) and layers on settings stored inside the iModel itself. iTwin-level settings are not loaded automatically — see [Referencing iTwin settings from an iModel](#referencing-itwin-settings-from-an-imodel) for how to include them. Because iModel-level dictionaries are loaded at [SettingsPriority.iModel]($backend) — the highest built-in priority — they override any lower-priority setting with the same name.

Use iModel settings when a particular iModel needs configuration that differs from the rest of its iTwin, or when you want to persist metadata (like an iTwin settings container reference) inside the iModel so it is available in future sessions.

### Persisted vs session-only dictionaries

There are two ways to supply settings dictionaries to an iModel's workspace scope:

| Method | Scope | Persistence |
|--------|-------|-------------|

> | [Settings.addDictionary]($backend) | Current session only | Values exist in memory only and are lost when the iModel is closed. |

Reviewer 2: you already mentioned how addDictionary is session-only in `## Settings dictionaries`, this seems repetitive

Author: Removed — already covered in the Settings dictionaries section.

| [EditTxn.saveSettingDictionary]($backend) (via `withEditTxn`) | All future sessions | Values are written to the iModel's `be_Props` table and automatically reloaded every time the iModel is opened. |

> Use `addDictionary` for transient overrides — for example, to inject ephemeral configuration while a particular tool is active. Use `saveSettingDictionary` only if you need settings to persist as part of an iModel's permanent record and cannot use a settings container instead.

Reviewer 2: same here

Author: Removed along with the table above.

### Saving iModel settings

To save settings into an iModel, call [EditTxn.saveSettingDictionary]($backend) (via `withEditTxn`) with a dictionary name and a [SettingsContainer]($backend) of key-value pairs:

```ts
[[include:WorkspaceExamples.saveSettingDictionary]]
```

The dictionary name (e.g. `"landscapePro/iModelSettings"`) identifies the dictionary within the iModel. If a dictionary with that name already exists, it is replaced; otherwise a new one is created. You can save multiple dictionaries under different names.

Note that modifying iModel settings requires an exclusive write lock on the entire iModel (obtained automatically by `withEditTxn`) — ordinary users should never do this.

### Deleting iModel settings

To remove an entire settings dictionary from an iModel, use [EditTxn.deleteSettingDictionary]($backend) (via `withEditTxn`):

```ts

> await withEditTxn(iModel, async (txn) => txn.deleteSettingDictionary("landscapePro/iModelSettings"));

Reviewer 2: how come this isn't in WorkspaceExamples?

Author: Moved to an extract snippet: `WorkspaceExamples.DeleteIModelSettingDictionary`.

### Referencing iTwin settings from an iModel

An iModel doesn't inherently know which iTwin settings container it should use. By saving the container's identity as an iModel-level setting, the iTwin settings become part of the iModel's workspace — so [IModelDb.workspace]($backend) becomes the single place to resolve all settings and resources the iModel uses.

```ts
[[include:WorkspaceExamples.SaveITwinSettingsReferenceInIModel]]
```

The next time the iModel is opened, your application reads the saved reference and passes it to the [IModelHost.getITwinWorkspace]($backend) overload that accepts [WorkspaceDbSettingsProps]($backend):

```ts
const settingsRef = iModel.workspace.settings.getSetting<WorkspaceDbSettingsProps>("landscapePro/itwinSettingsRef");
if (settingsRef !== undefined) {
  const iTwinWorkspace = await IModelHost.getITwinWorkspace(settingsRef);
  // iTwinWorkspace.settings now contains the iTwin-level settings.
  // ... use the settings, then close when finished:
  iTwinWorkspace.close();
}

> ```

Reviewer 2: should this be in workspace examples?

Author: The nearby inline snippet for loading settings from an iModel has been moved to an extract: `WorkspaceExamples.LoadITwinSettingsFromIModel`.

By default this resolves to the latest version of those settings.

### Pinning iTwin settings versions

If you need to pin the iModel to a specific version of the iTwin settings — so that its configuration does not change even when the iTwin settings are updated — save the version alongside the container props:

```ts
[[include:WorkspaceExamples.VersionAndPinITwinSettings]]
```

> ## Settings containers

Reviewer 2: This makes me think we should remove the "1 settings container per iTwin" restriction if we are encouraging people to make their own settings containers with non-standard attributes (metadata, manifest, etc.) since there can only be one of them.

Author: I've moved the admin workflows (discover, create, update, version) into a separate "Advanced: managing settings containers" section, framed generically rather than iTwin-specific. The conceptual intro stays under "Settings containers" with a note that getITwinWorkspace handles the common case automatically.
 
 I think we should document that these APIs are generic, and if needed, can be used to create multiple settings containers. They might not be iTwin level settings containers in the cloud, but something else. Domain rules and constraints of 1 settings container per iTwin is a rule we shouldn't enforce across the whole concept of settings....

Reviewer 2: > They might not be iTwin level settings containers in the cloud, but something else. Domain rules and constraints of 1 settings container per iTwin is a rule we shouldn't enforce across the whole concept of settings....

The thing is, you always create a container associated to an iTwin, its a required part of the API. And we throw an error when calling `getITwinContainerId` if there is more than 1 settings container associated to the iTwin. We did it like that for a reason - given an iTwinId, I need to find the settings container, if there's more than one because some admin somewhere created a bunch, what do we do? Its the same problem as before... I suppose you could just say load all settings containers associated to an iTwin and treat them collectively as the "settings"... but the code does not work like that today.

Today, if a user follows this tutorial and creates a new settings container in an iTwinId after someone has already created a settings container using `getITwinWorkspace` in the same iTwin, `getITwinWorkspace` will no longer work because of the constraint.

Author: > Today, if a user follows this tutorial and creates a new settings container in an iTwinId after someone has already created a settings container using `getITwinWorkspace` in the same iTwin, `getITwinWorkspace` will no longer work because of the constraint.

Small clarification on the creation path: the first default settings container comes from the write side via [`IModelHost.saveSettingDictionary`](https://github.com/iTwin/itwinjs-core/blob/nam/settings-workspace-docs/core/backend/src/IModelHost.ts#L493-L500), which goes through [`SettingsEditor.constructForITwin`](https://github.com/iTwin/itwinjs-core/blob/nam/settings-workspace-docs/core/backend/src/internal/workspace/SettingsEditorImpl.ts#L53-L69) and falls through to [`createNewCloudContainer(..., containerType: "settings")`](https://github.com/iTwin/itwinjs-core/blob/nam/settings-workspace-docs/core/backend/src/internal/workspace/WorkspaceImpl.ts#L520-L525) when no existing container is found. [`IModelHost.getITwinWorkspace(iTwinId)`](https://github.com/iTwin/itwinjs-core/blob/nam/settings-workspace-docs/core/backend/src/IModelHost.ts#L458-L465) is the read/discovery side via [`SettingsContainers.getITwinSettingsSources`](https://github.com/iTwin/itwinjs-core/blob/nam/settings-workspace-docs/core/backend/src/workspace/SettingsEditor.ts#L100-L123). So I think the “created using `getITwinWorkspace`” part is a little off, but your broader point about multiple settings containers breaking automatic selection still stands; that failure mode is already covered in [ITwinWorkspace.test.ts](https://github.com/iTwin/itwinjs-core/blob/nam/settings-workspace-docs/core/backend/src/test/standalone/ITwinWorkspace.test.ts#L114-L121).

I’ve updated the docs/snippets to align with that. The overview now calls out the read/write split in [WorkspacesAndSettings.md](https://github.com/iTwin/itwinjs-core/blob/nam/settings-workspace-docs/docs/learning/backend/WorkspacesAndSettings.md#L12-L18), the tutorial now says first creation happens on first write in [WorkspaceTutorial.md](https://github.com/iTwin/itwinjs-core/blob/nam/settings-workspace-docs/docs/learning/backend/WorkspaceTutorial.md#L84-L99), and the reference/docs/snippet now make the default single-container assumption explicit in [Settings.md](https://github.com/iTwin/itwinjs-core/blob/nam/settings-workspace-docs/docs/learning/backend/Settings.md#L157-L186), [Settings.md](https://github.com/iTwin/itwinjs-core/blob/nam/settings-workspace-docs/docs/learning/backend/Settings.md#L280-L330), and [WorkspaceExamples.test.ts](https://github.com/iTwin/itwinjs-core/blob/nam/settings-workspace-docs/example-code/snippets/src/backend/WorkspaceExamples.test.ts#L637-L643).

> > **When do you need this section?** The [IModelHost.saveSettingDictionary]($backend) and [IModelHost.deleteSettingDictionary]($backend) APIs shown in [iTwin settings](#itwin-settings) are sufficient for most application code — they manage the underlying settings container automatically. The `WorkspaceEditor`-based workflow described here is for **administrators** who need fine-grained control: creating containers explicitly, versioning, inspecting individual resources, or updating specific settings without replacing an entire dictionary.

Reviewer 2: I don't follow - what use-case would someone actually need to create a container themselves? Remember, getITwinWorkspace will create a settings container for you, and you can use it to inspect resources, update specific settings, etc.

Author: I added an advanced section now has a clear callout that most apps don't need this, and points to 
 getITwinWorkspace. I do want to keep the part on discovering, updating, creating containers available, docs extensibility sake (and openness)
